File: docs/docs/src/node/user-guide/concepts/contexts.js
// --8<-- [start:setup]
const pl = require("nodejs-polars");
const Chance = require("chance");

const chance = new Chance(42);
// --8<-- [end:setup]

// --8<-- [start:dataframe]

const arr = Array.from({ length: 5 }).map((_) =>
  chance.floating({ min: 0, max: 1 }),
);

let df = pl.DataFrame({
  nrs: [1, 2, 3, null, 5],
  names: ["foo", "ham", "spam", "egg", null],
  random: arr,
  groups: ["A", "A", "B", "C", "B"],
});
console.log(df);
// --8<-- [end:dataframe]

// --8<-- [start:select]
let out = df.select(
  pl.col("nrs").sum(),
  pl.col("names").sort(),
  pl.col("names").first().alias("first name"),
  pl.mean("nrs").multiplyBy(10).alias("10xnrs"),
);
console.log(out);
// --8<-- [end:select]

// --8<-- [start:filter]
out = df.filter(pl.col("nrs").gt(2));
console.log(out);
// --8<-- [end:filter]

// --8<-- [start:with_columns]

df = df.withColumns(
  pl.col("nrs").sum().alias("nrs_sum"),
  pl.col("random").count().alias("count"),
);

console.log(df);
// --8<-- [end:with_columns]

// --8<-- [start:groupby]
out = df.groupBy("groups").agg(
  pl
    .col("nrs")
    .sum(), // sum nrs by groups
  pl
    .col("random")
    .count()
    .alias("count"), // count group members
  // sum random where name != null
  pl
    .col("random")
    .filter(pl.col("names").isNotNull())
    .sum()
    .suffix("_sum"),
  pl.col("names").reverse().alias("reversed names"),
);
console.log(out);
// --8<-- [end:groupby]

File: docs/docs/src/rust/user-guide/concepts/contexts.rs
use polars::prelude::*;


fn main() -> Result<(), Box<dyn std::error::Error>> {
    
    // --8<-- [start:dataframe]
    use rand::{thread_rng, Rng};
    
    let mut arr = [0f64; 5];
    thread_rng().fill(&mut arr);

    let df = df! (
        "nrs" => &[Some(1), Some(2), Some(3), None, Some(5)],
        "names" => &[Some("foo"), Some("ham"), Some("spam"), Some("eggs"), None],
        "random" => &arr,
        "groups" => &["A", "A", "B", "C", "B"],
    )?;

    println!("{}", &df);
    // --8<-- [end:dataframe]

// --8<-- [start:select]
    let out = df
        .clone()
        .lazy()
        .select([
            sum("nrs"),
            col("names").sort(false),
            col("names").first().alias("first name"),
            (mean("nrs") * lit(10)).alias("10xnrs"),
        ])
        .collect()?;
    println!("{}", out);
// --8<-- [end:select]

// --8<-- [start:filter]
    let out = df.clone().lazy().filter(col("nrs").gt(lit(2))).collect()?;
    println!("{}", out);
// --8<-- [end:filter]

// --8<-- [start:with_columns]
    let out = df
        .clone()
        .lazy()
        .with_columns([
            sum("nrs").alias("nrs_sum"),
            col("random").count().alias("count"),
        ])
        .collect()?;
    println!("{}", out);
// --8<-- [end:with_columns]

    // --8<-- [start:groupby]
    let out = df
        .lazy()
        .groupby([col("groups")])
        .agg([
            sum("nrs"),                           // sum nrs by groups
            col("random").count().alias("count"), // count group members
            // sum random where name != null
            col("random")
                .filter(col("names").is_not_null())
                .sum()
                .suffix("_sum"),
            col("names").reverse().alias("reversed names"),
        ])
        .collect()?;
    println!("{}", out);
    // --8<-- [end:groupby]
    Ok(())
}